docs/CHAT_ADAPTER.md

```diff
@@ -0,0 +1,110 @@
+# How to implement a chat adapter?
+
+You can write your own chat adapter to use Web Chat on other services.
+
+Currently, the chat adapter interface Web Chat is using is adopted from the [`BotFramework-DirectLineJS`](https://github.com/microsoft/BotFramework-DirectLineJS) (DLJS) project. This is a "streaming" interface, the chat adapter do not keep any history or state about activities.
+
+The DLJS interface heavily relies on [Observable](https://github.com/tc39/proposal-observable) and [Bot Framework REST API](https://docs.microsoft.com/en-us/azure/bot-service/rest-api/bot-framework-rest-connector-api-reference?view=azure-bot-service-4.0#activity-object).
+
+## `activity$: Observable<Activity>`
+
+An Observable to notify Web Chat when an [Activity](https://docs.microsoft.com/en-us/azure/bot-service/rest-api/bot-framework-rest-connector-api-reference?view=azure-bot-service-4.0#activity-object) arrives. This Observable must contains all recent activities in the conversation, including both incoming and outgoing activities.
+
+When Web Chat post an outgoing activity, this Observable should notify Web Chat with that activity. And the service is expected to "fill-in-the-blank" for the outgoing activity, such as timestamp (the time when the activity is being posted to the service), service-assigned ID, sender ID, conversation ID, etc.
+
+Activity with `from.role` of `"user"` means it is an activity coming from the user who is using Web Chat. Activity with `from.role` of `"bot"` are from other users, including bots.
+
+### Updating activities
+
+Web Chat uses 2 different IDs to identify if 2 activities are the same or not:
+
+-  `activity.channelData.clientActivityID` is an ID set by Web Chat on all outgoing activity from the current session, ignored if the value is falsy
+-  `activity.id` is an ID set by the chat platform
+
+If either one of the IDs are the same, Web Chat will consider both are identifying the same activity. And it will replace the activity with last writer win strategy.
```

Comment:
```suggestion
If either one of the IDs are the same, Web Chat will conclude that both are identifying the same activity, and it will replace the activity with 'last writer wins' strategy.
```